dartagnan/src/main/java/com/dat3m/dartagnan/verification/solving/RefinementSolver.java

```diff
@@ -605,13 +608,54 @@ private boolean process(Constraint c) {
     }
 
     /*
-        This pass rewrites the memory model so that no positive constraints appear on the rhs-side of a difference.
-        Suppose "r = a \ b" where r and a are negative and b is positive.
+        This pass rewrites the memory model so that no non-negative constraints appear on the RHS of a difference.
+        Consider "r = a \ b" where b is non-negative (note that r and a must be non-positive then).
         For example, this is the case if there is a constraint "~empty(r)"
         We rewrite the model to
-            let b`= free()
+            let b'= free()
             let r = a \ b' // b' gets cut, but it is effectively a base relation so this causes no large encoding.
-            empty (r & b) // b is now in a positive position and thus not need to get cut
+            empty (r & b) // b is now in a positive position and thus not need to get cut.
+
+        We claim that this model is equisatisfiable to the original memory model.
+        To understand this intuitively, it is helpful so think of the fresh b' as the original b but with "some slack".
+        We will argue that the best choice for b' is precisely b.
+        Any larger choice for b' will make r smaller, but since r is non-positive this reduces the chances
+        of satisfying negated axioms (e.g., if "~empty(r)" is true for some small r, then it is also true for all larger r).
+        On the other hand, if b' is chosen too small, then r gets so large that it violates "empty(r & b)".
+        We prove this formally:
+
+        "NEW => ORIGINAL":
+        Suppose M' is a consistent execution of the new memory model.
+        Then from "let r = a \ b'" it follows that M'(r) \subseteq M'(a).
```

Comment:
Is `M'(r)` the restriction of `r` to the pairs of executed events in `M'`?

Reply:
No, its the interpretation of `r` in the given execution `M'`. `r` can take many values depending on the concrete execution (i.e., depending on the values taken by the base relations like `rf/co`). Without an execution `M'`, `r` is just a syntactic element / a variable.